src/app/gravatar/gravatar.service.ts

```diff
@@ -0,0 +1,24 @@
+import { Injectable } from '@angular/core';
+import { Md5 } from 'ts-md5/dist/md5';
+
+@Injectable({
+  providedIn: 'root',
+})
+export class GravatarService {
+  readonly gravatarBaseUrl = 'https://www.gravatar.com/avatar/';
+  constructor() {}
+
+  public gravatarUrlForEmail(email: string, defaultImage: string) {
+    const hash = Md5.hashStr(email);
+    return `${this.gravatarBaseUrl}${hash}}?id=${defaultImage}&s=500`;
+  }
+
+  public gravatarUrlForMysteryPerson() {
+    // https://en.gravatar.com/site/implement/images/ -- it's "mp", not "mm" there, but we've had it like this
+    return `${this.gravatarBaseUrl}?d=mm&s=500`;
+  }
+
+  public gravatarUrlForImageUrl(imageUrl: string | null) {
+    return imageUrl ? `${this.gravatarBaseUrl}000?d=${imageUrl}` : null;
```

Comment:
We should urlencode any uncontrolled query string values here and elsewhere via `encodeURIComponent`, so that a value with sensitive characters (for example, an `imageUrl` containing `&` ala `http://www.site.com/images?p=foo&q=bar`) gets through correctly.

Reply:
Good catch; we'd been doing it this way all along, and surprisingly it worked. I guess the browser was encoding it. Changed and fixed in 22934034d46cae1dafc121d86bd77b5d771c6c53